LSPS1/extensions/refunds/README.md

```diff
@@ -0,0 +1,36 @@
+# LSPS1.refunds
+
+| Name    	| refunds                                     	|
+|---------	|------------------------------------------------	|
+| Version 	| 1                                              	|
+| OpenApi 	| [LSPS1.refunds.yaml](./LSPS1.refunds.yaml) 	|
+
+
+
+## API
+
+The extension `refunds` adds two endpoints to automatically manage refunds.
```

Comment:
Refunds are not strictly needed if LSP is only accepting Lightning payments, as refusing to release the preimage is automatically a refund.  But if you are going to use onchain payments at all, it needs refunds unless you want to do onchain contracting of some kind.

Reply:
Agree. I worked on the spec the last days and I incorporated HOLD invoices (see new commits). Onchain needs another refund mechanism (todo).

Reply:
Why hold it though?  Just refuse it outright and give the error reason in the new fat error mechanism.

Reply:
Normal invoices (cln, lnd, not idea about LDK) release the preimage instantly when the payment arrives. You need to hold the payment for the duration of the channel open process at least.

1. Payment arrives at the LSP.
2. LSP holds payment.
3. LSP tries to open the channel.
4. LSP resolves/rejects payment depending on the channel open success.

Reply:
What I meant is that as soon as LSP *allocates* money for it, claim the payment.  If LSP cannot allocate money, reject it.  No need to wait for network connection or channel opening or confirmation of channel.

This can be implemented in Core Lightning via `htlc_accepted` plugin hook so that the plugin can wait for money allocation procedure, and `signinvoice` command to create a non-automatic invoice.  I assume LND has similar since peerswap is implemented for it and peerswap would (probably?) use something similar, or maybe not.  LDK still needs some hooks on payment release I think.

Reply:
I need to specify what error the LSP should set in case it doesnt have funds. The LSP SHOULD already throw this error at the order creation call `POST /lsp/channels`. It MUST set the error after the channel open failed.

I am unsure what you mean with the "money allocation" step. In the case you open every channel with it's own transaction, the money is allocated right before the funding transaction is published. The payment SHOULD be claimed after the funding transaction is published. I don't see a DDoS vector here.

Are you referring to the batched channel open case? Here I can see that the steps `money allocation` and `funding transaction published` could be different. I have not thought through this case completely yet though.

Reply:
>The LSP SHOULD already throw this error at the order creation call `POST /lsp/channels`

If you reserve funds before getting paid, you are left vulnerable to someone spamming `POST /lsp/channels` on you and preventing you from allocating funds for other clients.  You have to reserve funds at the point you get paid, not at `POST` time.

>I am unsure what you mean with the "money allocation" step

Situation is this:

* LSP has 1 BTC.
* Client A wants 1 BTC of capacity
* Client B wants 1 BTC of capacity
* Client A makes a `POST` request.
* Client B makes a `POST` request.
  *  If LSP errors at this point because LSP already allocated 1.0 BTC to client A, client A can simply not pay the invoice, thus keeping the 1 BTC reserved for client A, and any future client Bs are unable to acquire service, i.e. denial of service.
* Client B tries to pay the invoice.  LSP claims the payment and reserves 1.0 BTC to client B, waiting for client B to connect or ping the LNURL (why tho) or whatever.
* Client A tries to pay the invoice.
  * This is the proper place to report the error and fail to claim the invoice payment, and is what I am referring to.

Basically you have to allocate funds, at some point, and that allocation can fail because nobody has infinite Bitcoins, it is literally just 21 million.  If you error early, before getting paid, as noted above client A can just reserve your funds.

Allocation need not be tied to acutally publishing funding transaction.  Presumably the LSP knows the value of every UTXO it controls, and can easily sum them up and check if the reserved funds is less than or equal to that value.